### PR TITLE
Fixed #29916 -- Added lower_inc, lower_inf, upper_inc, and upper_inf lookups for RangeFields.

### PR DESCRIPTION
**Pull request by dulmandakh**

### CODE REVIEW

File: django/contrib/postgres/fields/ranges.py [hunk 1]
```diff
@@ -159,6 +159,14 @@ def db_type(self, connection):
...
 RangeField.register_lookup(lookups.ContainedBy)
 RangeField.register_lookup(lookups.Overlap)
+DateRangeField.register_lookup(lookups.LowerInclusive)
+DateRangeField.register_lookup(lookups.LowerInfinite)
+DateRangeField.register_lookup(lookups.UpperInclusive)
+DateRangeField.register_lookup(lookups.UpperInclusive)
+DateTimeRangeField.register_lookup(lookups.LowerInclusive)
+DateTimeRangeField.register_lookup(lookups.LowerInfinite)
+DateTimeRangeField.register_lookup(lookups.UpperInclusive)
+DateTimeRangeField.register_lookup(lookups.UpperInclusive)
```
Comment (felixxm) [inline django/contrib/postgres/fields/ranges.py]: These lookups should be available for all `RangeField`'s not only for `DateRangeField` and `DateTimeRangeField`.

File: django/contrib/postgres/lookups.py [hunk 2]
```diff
@@ -46,12 +46,36 @@ class HasAnyKeys(HasKeys):
...
+    output_field = BooleanField()
+
+
 class Unaccent(Transform):
     bilateral = True
     lookup_name = 'unaccent'
     function = 'UNACCENT'
 
 
+class UpperInclusive(Transform):
```
Comment (felixxm) [inline django/contrib/postgres/lookups.py]: Please use `@RangeField.register_lookup` and move these transforms to `django/contrib/postgres/fields/ranges.py`.

File: tests/postgres_tests/test_ranges.py [hunk 3]
```diff
@@ -283,6 +283,46 @@ def test_date_range(self):
...
+
+    def test_date_range_upper_inf(self):
+        objs = [
+            RangesModel.objects.create(dates=DateRange('2015-01-01', '2015-05-04')),
+            RangesModel.objects.create(dates=DateRange('2015-05-04', None)),
+        ]
+        self.assertSequenceEqual(
+            RangesModel.objects.filter(dates__upper_inf=True),
+            [objs[1]],
+        )
```
Comment (felixxm) [inline tests/postgres_tests/test_ranges.py]: Please move these tests to the `TestQuerying`. Try to reuse existing objects (see `TestQuerying.setUpTestData()`, it should be enough to add one new object with an infinite upper bound, e.g. `RangesModel(ints=NumericRange(1, None))`,

File: django/contrib/postgres/lookups.py (hunk 2 above)
Comment (dulmandakh) [reply django/contrib/postgres/lookups.py]: got it

File: tests/postgres_tests/test_ranges.py (hunk 3 above)
Comment (dulmandakh) [reply tests/postgres_tests/test_ranges.py]: ok

**Comment by dulmandakh** [comment]
@felixxm is it possible to cherry-pick this change into 3.0? It would be great if possible. Thanks

**Comment by felixxm** [comment]
@dulmandakh No, Django 3.0 is already after a features freeze (see https://code.djangoproject.com/wiki/Version3.0Roadmap).

**Comment by dulmandakh** [comment]
so sorry, I'll maintain my own copy of the feature until 3.1 👍

File: docs/releases/3.0.txt [hunk 4]
```diff
@@ -169,6 +169,12 @@ Minor features
 * The new :class:`~django.contrib.postgres.fields.RangeBoundary` expression
   represents the range boundaries.
 
+* The new :class:`~django.contrib.postgres.ranges.LowerInclusive`,
```
Comment (pope1ni) [inline docs/releases/3.0.txt]: Please move this to `docs/releases/3.1.txt`.

File: django/contrib/postgres/fields/ranges.py [hunk 5]
```diff
@@ -259,6 +259,34 @@ class AdjacentToLookup(lookups.PostgresSimpleLookup):
...
+    lookup_name = 'upper_inc'
+    function = 'UPPER_INC'
+    output_field = models.BooleanField()
+
+
+@RangeField.register_lookup
+class UpperInfinite(models.Transform):
+    lookup_name = 'upper_inf'
+    function = 'UPPER_INF'
+    output_field = models.BooleanField()
```
Comment (pope1ni) [inline django/contrib/postgres/fields/ranges.py]: These need to be documented in `docs/ref/contrib/postgres/fields.txt`.

**Comment by dulmandakh** [comment]
@pope1ni removed docs changes, please help with documentation. Thank you

**Comment by dulmandakh** [comment]
@felixxm managed to setup local environment, and all tests are passing

**Comment by dulmandakh** [comment]
exporting LowerInclusive, LowerInfinite, UpperInclusive, UpperInfinite, IsEmpty because these can be used with CheckConstraint, or it was my main motivation.

**Comment by felixxm** [comment]
@dulmandakh Thanks! I added docs and release notes, rebased from master, squashed commits and pushed minor edits.

**Comment by dulmandakh** [comment]
@felixxm @carltongibson thank you for the review. I'm thinking about other contributions/improvements to Django in the future.

**Comment by felixxm** [comment]
I switched tests to `RangesModel.decimals` because `ints` are represented by `int4range` (see [Discrete Range Types](https://www.postgresql.org/docs/9.6/rangetypes.html#RANGETYPES-DISCRETE)), so we will not be able to test all combinations.

**Comment by felixxm** [comment]
@carltongibson Thanks :+1: I pushed small amendments.

**Pull request merged**